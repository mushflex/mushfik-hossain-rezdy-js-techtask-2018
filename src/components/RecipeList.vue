<template>
	<div class="recipe-list">
		<div class="loading" v-if="loading">
			Loading...
		</div>

		<div v-if="error" class="error">
			{{ error }}
		</div>

		<div v-if="recipeData" class="content">
			<h4>recipe list:</h4>
			<ul>
				<li>{{ recipeData.name }}</li>
			</ul>
		</div>
	</div>
</template>

<script>
import _ from 'lodash'

export default {
	name: 'RecipeList',
	props: [ 'buttonLabel' ],
	data () {
		return {
			loading: false,
			recipes: null,
			ingredients: null,
			error: null
		}
	},
	async created () {
		// fetch the data when the view is created and the data is
		// already being observed
		const allRecipes = [];
		const allIngredients = [];
			
		await fetch('static/ingredients.json')
      .then(ingredients => ingredients.json())
      .then(json => {
				this.ingredients = this.filterIngredients(json.ingredients);
			});
		console.log(this.ingredients);

		await fetch('static/recipes.json')
      .then(r => r.json())
      .then(json => {
				this.recipes = this.filterRecipes(json.recipes);
			});

			console.log(this.recipes)
	},
	methods: {
		filterIngredients: function (ingredients) {
			console.log(ingredients)
			const today = new Date();
			const oldIngredients = _.map(ingredients, (ingredient) => {
				const useBy = new Date(ingredient['use-by']);
				const bestBefore = new Date(ingredient['best-before']);
				// ingredient is before the useBy date
				if ((today < useBy) && (today > bestBefore)) {
					return ingredient;
				} 
			});
			
			// console.log(oldIngredients)

			const freshIngredients = _.map(ingredients, (ingredient) => {
				const bestBefore = new Date(ingredient['best-before']);
				// ingredient is before the best before date
				if (today < bestBefore) {
					return ingredient;
				}
			});
			

			// console.log(freshIngredients)
			
			return {
				oldIngredients: _.pickBy(oldIngredients, _.identity),
				freshIngredients: _.pickBy(freshIngredients, _.identity)
			};
		},

		filterRecipes: function (recipes) {
			const recomendedRecipes = recipes.map(recipe => {
				return recipe;
			});

			return {
				topRecipeList: [],
				bottomRecipeList: []
			}
		},
		goBack: function () {
			this.$router.push('/');
		}
	}
}
</script>

<!-- Add "scoped" attribute to limit CSS to this component only -->
<style scoped>
</style>
