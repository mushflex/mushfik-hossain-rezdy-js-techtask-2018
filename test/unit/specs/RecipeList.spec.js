import Vue from 'vue'
import { shallow } from 'vue-test-utils'
import RecipeList from '@/components/RecipeList'

describe('RecipeList.vue', () => {
	it('has the expected html structure', () => {
		const Constructor = Vue.extend(RecipeList);
		const vm = new Constructor().$mount();
		expect(vm.$el).toMatchSnapshot();
	})

	it('should render header correctly', () => {
		const Constructor = Vue.extend(RecipeList);
		const vm = new Constructor().$mount();
		const header = 'Recipe List';
		expect(vm.$el.querySelector('h1').textContent).toEqual(header);
	})

	it('should render loading when loading is true', () => {
		const Constructor = Vue.extend(RecipeList);
		const vm = new Constructor().$mount();
		expect(vm.loading).toEqual(true);
	})

	it('should render empty content when not loading', () => {
		const cmp = shallow(RecipeList, { // Create a shallow instance of the component
			data: {
				loading: false,
				recipes: null,
				ingredients: null
			}
		})
		expect(cmp.vm.loading).toEqual(false);
		expect(cmp.vm.$el.querySelector('.content').textContent).toEqual('');
	})

	it('should render empty content when not loading', () => {
		const cmp = shallow(RecipeList, { // Create a shallow instance of the component
			data: {
				loading: false,
				recipes: {
					topRecipeList: [ { title: 'tiramisu' } ],
					bottomRecipeList: [ { title: 'green salad' } ]
				},
				ingredients: null
			}
		})
		expect(cmp.vm.$el.querySelector('.recipe-list__list').textContent).toEqual('tiramisu green salad');
	})
})
