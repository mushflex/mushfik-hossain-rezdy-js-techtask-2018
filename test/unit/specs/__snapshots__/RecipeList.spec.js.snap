// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecipeList.vue has the expected html structure 1`] = `
<div
  class="recipe-list"
>
  <h1>
    Recipe List
  </h1>
   
  <div
    class="loading"
  >
    
		Loading...
	
  </div>
   
  <!---->
</div>
`;
